---
title: "Hands On Session: exploring microhaplotypes with 'microhaplot'"
output: html_notebook
author: Eric C. Anderson
---


The R-package program `microhaplot` provides functionality to extract 
microhaplotypes from aligned reads in .sam files, and to display
information about reads containing microhaplotypes in a variety of 
useful visualizations.  The goals is to give users the chance to dig
into the patterns in their data and assess the quality of data at different
microhaplotype loci.

## Installation

This R package can be installed from GitHub:
```{r}
devtools::install_github("ngthomas/microhaplot", build_vignettes = TRUE)
```

### Known issues

* If you are working on a mac and just upgraded to a new major version of 
R, you should also reinstall [XQuartz](https://www.xquartz.org/). 
* If you are working on windows and are using R 3.5.1, things won't work
out because `microhaplot` makes use of the `feather` package, which is currently
not built for 3.5.1.  :-(.

## After installing some steps to launch it

You have to load the package, move the Shiny app materials to a good location
and then launch the Shiny app:
```{r}
# provide a directory path to host the haPLOType app
microhaplot::mvHaplotype("~/Shiny")

# launch the Shiny app:
library(microhaplot)
app.path <- "~/Shiny/microhaplot"
runHaplotype(app.path)

```


Now, you can follow along with the first vignette:
```{r}
vignette("haPLOType-walkthrough")
```

If you want more background on preparing your data for microhaplot try this:
```{r, eval=FALSE}
vignette("haPLOT-data-prep")
```


I will spend a little time showing a new feature that is mostly undocumented---the
"Scatter Analysis," which was suggested by Phil Morin, and is a great way of looking at 
the data.
